main/lsp/requests/document_formatting.cc

```diff
@@ -83,34 +86,33 @@ void DocumentFormattingTask::index(LSPIndexer &index) {
                 // Construct text edit to replace entire document.
                 // Note: VS Code uses 0-indexed lines, so the lineCount will be one more line than the size of the
                 // doc.
-                edits.emplace_back(
-                    make_unique<TextEdit>(make_unique<Range>(make_unique<Position>(0, 0),
-                                                             make_unique<Position>(index.getFile(fref).lineCount(), 0)),
-                                          formattedContents));
+                edits.emplace_back(make_unique<TextEdit>(
+                    make_unique<Range>(make_unique<Position>(0, 0), make_unique<Position>(originalLineCount, 0)),
+                    formattedContents));
                 result = move(edits);
                 response->result = move(result);
                 break;
             case RubyfmtStatus::SYNTAX_ERROR:
-                displayError(fmt::format("`rubyfmt` could not format {} because it contains syntax errors.",
-                                         index.getFile(fref).path()),
-                             response);
+                response->error = make_unique<ResponseError>(
+                    (int)LSPErrorCodes::RequestFailed,
+                    fmt::format("`rubyfmt` could not format {} because it contains syntax errors.", path));
```

Comment:
ooc, is this what other formatters do when there are syntax errors?

i wonder if it might be better to return the file contents unchanged?


> result: [TextEdit[]](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textEdit) | null describing the modification to the document to be formatted.
> error: code and message set in case an exception happens during the formatting request.

i wonder if maybe it expects a `null` response to be sent in that case, instead of an `error`?

Reply:
Fair enough, not sure about what other formatters do but I think it's reasonable to say that syntax errors are somewhat "expected" and can be no-ops.

Reply:
Actually, I just looked at the prettier logs for typescript, and it does look like it returns an error. I made a junk change and saw this:

```
["ERROR" - 10:36:43 AM] Error formatting document.
["ERROR" - 10:36:43 AM] ';' expected. (2:4)
  1 | export {};
> 2 | fn fn fn sdfsdlj
    |    ^
  3 | 
SyntaxError: ';' expected. (2:4)
  1 | export {};
> 2 | fn fn fn sdfsdlj
    |    ^
  3 | 
```

I don't know of an easy way to see whether they simply logged the error and returned `null` or if they actually returned an error type, but that's at least another data point for how to handle it. Going off of that, I'm inclined to return this as an error. Invalid syntax _is_ an error to `rubyfmt` after all, so I think reflecting that in the response type makes sense to me, but let me know if you feel otherwise.

Reply:
Here is where rust analyzer returns Ok(None) instead of Err(…) for syntax errors:

https://github.com/rust-lang/rust-analyzer/blob/7f9c054686fb3e1af7acb7292e05462afc138981/crates/rust-analyzer/src/handlers.rs#L1856

the comment in the source is worth reading too

Reply:
Interesting, I don't feel strongly either way, so happy to have it be `null` instead 👍

Reply:
Also I haven’t been able to find: does TypeScript’s LSP actually support textDocument/formatting requests? I thought that prettier was a separate VS Code plugin that used VS Code extension APIs, not the LSP spec for formatting.

unlike most of the time where we copy typescript when trying to resolve LSP ambiguities in sorbet, I don’t know whether typescript’s implementation is a good role model for formatting

Reply:
It does [apparently](https://github.com/typescript-language-server/typescript-language-server/blob/76bf9a4817ffa1e340422cfd5177dbcb96528ddb/src/lsp-server.ts#L717-L743)! I didn't look too terribly far into it, but it looks to me like it also swallows errors (of all kinds) and returns an empty `TextEdit[]` (and never `null`, interestingly).

Reply:
from the docs it seems like that is it’s “basic formatting”

https://code.visualstudio.com/docs/languages/typescript#_formatting

not prettier, so the error logs you saw from prettier were likely from the separate plugin

Reply:
Yeah, the prettier ones are definitely separate (and agreed that they're probably not the best example here).